src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -146,24 +147,16 @@ public bool HasValue
         /// </returns>
         public static int Compare(StringSegment a, StringSegment b, StringComparison comparisonType)
         {
-            int minLength = Math.Min(a.Length, b.Length);
-            int diff = string.Compare(a.Buffer, a.Offset, b.Buffer, b.Offset, minLength, comparisonType);
-            if (diff == 0)
-            {
-                diff = a.Length - b.Length;
-            }
+            // Use CompareInfo.Compare instead of MemoryExtensions.CompareTo because we need
+            // to distinguish between null and empty StringSegment values.
```

Comment:
Can't HasValue be used to discriminate between null and empty segments?

Reply:
Unless there are scenarios which I haven't considered the following passes all the StringSegment tests: https://github.com/bbartels/runtime/blob/bc223014a90472898906ed9a379e09e35dc62b7b/src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs#L151

Not sure about perf between `CompareInfo.Compare` and `ReadOnlySpan<char>.CompareTo`, though I assume CompareInfo.Compare is probably a wrapper around CompareTo?